content/blog/2022-04-27-terraform-provider.md

```diff
@@ -0,0 +1,118 @@
+
+The requirements for Machine Learning (ML) infrastructure are becoming
+increasingly complex. Training large models often requires specialized hardware
+(GPUs, TPUs) which involves moving the whole training process onto cloud
+machines, setting up environments and synchronizing data. For teams that want to
+leverage spot instances, the setup becomes even more complex - they need to make
+sure the training progress is not lost during spot instance recovery. This is
+time-consuming, and requires expertise in both DevOps and Machine Learning. On
+top of that, the costs of training in a cloud environment can be high due to not
+only the need for specific hardware, but also individual responsibility to shut
+down instances when training is complete.
+
+To address the specific needs of machine learning teams, we have built the
+Iterative Provider (TPI). TPI is an open-source tool extending the functionality
+of Terraform, the world's most widely used multi-cloud provisioning product. The
+Iterative Provider enables full lifecycle management of computing resources and
+is designed specifically for machine learning pipelines.
+
+## Tailored to Machine Learning Workflows
+
+The Iterative Provider offers a single resource called `iterative_task` which
+you can use to configure:
+
+- Your cloud infrastructure
+- The steps to perform on the cloud resource, i.e. setting up the environment,
+  running the training pipeline, logging metrics, etc.
+- The data to be synced back once the training is complete (e.g. a file with
+  metrics, a model, plots)
+
+Here’s a “hello world” example of a `main.tf` Terraform configuration file using
+the `iterative_task` resource:
+
+```hcl
+terraform {
+  required_providers { iterative = { source = "iterative/iterative" } }
+}
+provider "iterative" {}
+
+resource "iterative_task" "example" {
+  cloud   = "aws"
+  region  = "us-east-2"
+  machine = "l+k80"
+
+  storage {
+    workdir = "."
+    output  = "results"
+  }
+  script = <<-END
+    #!/bin/bash
+    sudo apt update
+    sudo apt install -y python3-pip
+    pip3 install --user -r requirements.txt
+    python3 train.py
+  END
+}
+```
+
+Once the training is complete, the Iterative Provider terminates the resource,
+so users don't have to worry about spiraling costs from unused machines.
+
+## Configure Once, Bring Everywhere
+
+Once you configure infrastructure and a script that executes your training
+pipeline in a Terraform configuration file, you can bring that pipeline anywhere
+you want. You can use such a config for ad-hoc training at any stage of your
+prototyping process or use it as a job in your preferred CI/CD tool. It will
+also store your infrastructure configuration files in a version control system
+together with the rest of your project for easier control.
+
+## One Provider to Rule Them All
+
+Whether you prefer Amazon Web Services (AWS), Microsoft Azure, Google Cloud
+Platform (GCP), or Kubernetes, the Iterative Provider has you covered. You can
```

Comment:
```suggestion
Platform (GCP), or Kubernetes (K8s), the Iterative Provider has you covered. You can
```

Reply:
> _One Provider to Rule Them All_

Clever for #devrel, “planeworthy” by and for #engineering. 💍

Reply:
One Provider to Rule Them All,
One Provider to `spot` them,
One ring to `apply` them all,
and in the darkness `destroy` them.